### PR TITLE
ROX-26344: adds runtime configuration to collector

### PR DESCRIPTION
**Pull request by Stringy**
## Description

Includes support for enabling/disabling external IPs based on runtime config, deferring to existing feature flag if none is provided

In subsequent PRs, the runtime configuration will be populated. 

## Checklist
- [x] Investigated and inspected CI test results
~- [x] Updated documentation accordingly~

**Automated testing**
  - [x] Added unit tests
  ~- [ ] Added integration tests~
  ~- [ ] Added regression tests~

If any of these don't apply, please comment below.

## Testing Performed

Unit tests added to exercise the intended behaviour. It is not possible to integration-test this until Sensor->Collector communication has been built.

### CODE REVIEW

**Review by ovalenti — COMMENTED** [review]
Using the CollectorConfig object makes sense, good idea !

I think that the next steps should be:
- on the Sensor side:
  - Update the [proto defs](https://github.com/stackrox/stackrox/blob/7b6bc042b9352869ccfbc5e928889cdc61ece7b5/proto/internalapi/sensor/network_connection_iservice.proto) to introduce an enum type wrapping and replacing `NetworkFlowsControlMessage`
  - introduce the runtime-config object in the new enum
  - implement a switch-case in NetworkStatusNotifier in [onRecvControlMessage](https://github.com/stackrox/collector/blob/giles/ROX-26344-integrate-runtime-config/collector/lib/NetworkStatusNotifier.cpp#L58C29-L69) or its successor
  - when receiving a runtime-config in the new switch-case, update the value in CollectorConfig
- on the ConnTracker side:
  - everytime we start a scrape in either `NetworkStatusNotifier::RunSingle` or `NetworkStatusNotifier::RunSingleAfterglow`, set the ExtIP attribute using [`ConnectionTracker::EnableExternalIPs`](https://github.com/stackrox/collector/blob/giles/ROX-26344-integrate-runtime-config/collector/lib/ConnTracker.h#L130C8-L130C25)

**Review by ovalenti — DISMISSED** [review]
As discussed previously and in [#11746 ](https://github.com/stackrox/stackrox/pull/11746), we should remember to unify the names of the environment variables that enable the ExtIP feature. 

https://github.com/stackrox/collector/blob/9828784f88b7b7f07b0af6b5af0653ce85af823a/collector/lib/CollectorConfig.cpp#L50

I think that it would make sense to update CollectorConfig's env var name to match the one in Central ([`ROX_EXTERNAL_IPS`](https://github.com/stackrox/stackrox/blob/ovalenti/learned_external_IPs/pkg/features/list.go#L133)) as part of this PR.

File: collector/lib/CollectorConfig.h
Comment (JoukoVirtanen) [inline collector/lib/CollectorConfig.h:85]: This function could be called `IsExternalIPsEnabled` or `GetExternalIPsEnabled` to make it clear that this checks the if external IPs should be enabled and it doesn't actually set any values. However, your naming is consistent with the rest of the code, so you can keep it.

File: collector/lib/CollectorConfig.h
Comment (Stringy) [reply collector/lib/CollectorConfig.h:89]: In this case, `network_connection_config()` returns a reference, so it can't be null

Comment (Molter73) [reply collector/lib/CollectorConfig.h:89]: In that case, could we change line 88 slightly to make it clearer?
```cpp
const auto& network_cfg = cfg.network_connection_config();
```

Also, `auto` on its own will create a copy of the object, which we might want to avoid here. So we probably want to change line 87 as well:
```cpp
const auto& cfg = runtime_config_.value();
```

https://stackoverflow.com/questions/29859796/c-auto-vs-auto
https://en.cppreference.com/w/cpp/language/auto

File: collector/lib/CollectorConfig.h
Comment (Molter73) [inline collector/lib/CollectorConfig.h:115]: Lets not copy the object every time it's accessed.
```suggestion
  const std::optional<sensor::CollectorConfig>& GetRuntimeConfig() const {
```

File: collector/lib/CollectorConfig.h
Comment (Molter73) [inline collector/lib/CollectorConfig.h:113]: I know we discussed this offline yesterday, but if we don't expect the caller to keep using the `runtime_config` object, we should probably just use an rvalue here:
```suggestion
  void SetRuntimeConfig(sensor::CollectorConfig&& runtime_config) {
    runtime_config_ = runtime_config;
  }
```
Alternatively, method overloading is a thing, we can have both and let the caller decide if they are `std::move`ing the object or not.

File: collector/lib/CollectorConfig.cpp
Comment (Molter73) [inline collector/lib/CollectorConfig.cpp:50]: I actually think the `ENABLE` makes it clearer what the env variable is for. But I would argue a variable in the form `<PROGRAM>_<FEATURE>_<PARAM>` is even more clear, so in this case we might want to name the variable `ROX_EXTERNAL_IPS_ENABLE` or `ROX_COLLECTOR_EXTERNAL_IPS_ENABLE`

File: collector/lib/CollectorConfig.h
Comment (Stringy) [reply collector/lib/CollectorConfig.h:89]: I didn't know that was a thing, thanks for the heads up!

File: collector/lib/CollectorConfig.cpp
Comment (Stringy) [reply collector/lib/CollectorConfig.cpp:50]: In general I think I agree. @ovalenti what do you think? should we update the central feature flag?

Comment (ovalenti) [reply collector/lib/CollectorConfig.cpp:50]: In Central, it looks like the convention is to use short names and no verb... https://github.com/stackrox/stackrox/blob/0b83c7607cb58d28e0165e20936f30d0c11d12c0/pkg/features/list.go

Although unifying the flag names seemed like a good idea in the first place, the collector one is likely not going to be used directly by any customer. So I am starting to think that we could keep the names as they were, or as Mauro suggests. If it was only my, the thing would not be named `EXTERNAL_IP` at all...

Comment (JoukoVirtanen) [reply collector/lib/CollectorConfig.cpp:50]: I don't think `ENABLE` is needed in the name. In stackrox/stackrox the function to check the value of feature flags is `Enable()`, so it would seem to be redundant.

File: collector/lib/CollectorConfig.h
Comment (JoukoVirtanen) [reply collector/lib/CollectorConfig.h:85]: Resolving as the naming here is consistent with the rest of the code.

**Review by JoukoVirtanen — APPROVED** [review]
Approving, but before merging update the stackrox/stackrox submodule as I merged the internal protobuf definitions. Also get approvals from others before merging.

File: collector/lib/CollectorConfig.cpp
Comment (Molter73) [reply collector/lib/CollectorConfig.cpp:50]: > In stackrox/stackrox the function to check the value of feature flags is `Enable()`, so it would seem to be redundant.

It wouldn't be redundant to a user that is trying to configure collector without any knowledge of what the code in the main repo looks like 🙂

Comment (Stringy) [reply collector/lib/CollectorConfig.cpp:50]: I will revert this for now and we can decide on naming conventions later on

**Pull request merged**
